### ISSUE TITLE
Consider creating an option to mimic previous tab behavior

### ISSUE THREAD

**Issue by non-Jedi** [JuliaEditorSupport/julia-emacs#193]
Per @ronisbr, #185 significantly change the behavior of <kbd>TAB</kbd>.

https://github.com/JuliaEditorSupport/julia-emacs/issues/188#issuecomment-1542914710:

> If I use the latest commit with `(setq julia-force-tab-complete t)` (default), I can replace LaTeX symbols by pressing TAB. For example, `"\alpha<TAB>"` gives `α`. However, in this case, TAB always force indentation update. Hence, I am not able, for example, to increase the indentation by pressing TAB repeatedly as I could before.
> 
> If I set `(setq julia-force-tab-complete nil)`, now I can increase the indentation by pressing TAB repeatedly. However, I lost the ability to replace LaTeX symbol by hitting TAB. In this case, I must use `C-SPC`, which is my default completion keystroke.
> 
> I am not saying we must revert to the old behavior. I am just saying that the behavior has changed and I am not certain if this is a bug or how the new system must work.

To me, this brings <kbd>TAB</kbd> behavior back in line with how users of the rest of emacs would expect it to behave, and I see it as a benefit. If there's sufficient interest, I could create a configuration option that would cause <kbd>TAB</kbd> to simulate the old behavior.

**Comment by ronisbr** [comment]
I really prefer the old behavior because it is consistent with REPL. If you hit TAB after a LaTeX symbol, it replaces it, and if you hit TAB at the beginning of the line, it increase indentation.

**Comment by tpapp** [comment]
@ronisbr: We can bring it back, and with the hash table lookup it should be easy, but let's focus on the details first.

In the examples below, `|` is the point. 

1. `\prefix|noise` matches multiple ones, eg `\prefix1`, `\prefix2`, etc. Here `noise` is irrelevant. What should it do, pick the first one, offer a list with a popup, or something else?
2. `\pre|fixnoise` matches `\pre` and `\prefix`, what should it expand to?

Note that the REPL behavior is the following on 1.9:

1. on `\al|eph`, it refuses to replace, just throws up `\alpha`, `\allequal`, and `\aleph`.
2. on `\al|noise`, same thing.
3. `\not|` will just expand greedily, and not offer `\notin`.
4. `\not|in` will expand to ` ̸in`, again greedily.

I think that the REPL is made for interactively entering text and expanding on whatever comes *before* the point. It is very rare (I think) to go back in the REPL prompt and expand text. Whereas in an editor there is text before and after the point.

I kind of missed the REPL behavior for a while, so I understand the issue, but in the end I arrived at the conclusion that editors need different semantics. I am very happy with the current setup now, as we get expansion *within* the symbol a way to resolve multiple options.

That said, I am open to a mechanism that mimics the REPL in some way. Just please spend some time thinking about the semantics.

**Comment by ronisbr** [comment]
Perfect @tpapp ! 

Maybe we should not really replicate the entire REPL behavior. Is there any technical problems to always call auto-completion if a symbol does not resolve to a single solution? So, in all those cases, it would open that popup if I hit TAB?

What I really miss is the ability to use TAB to increase indentation if we does not have a LaTeX symbol. For example, if I hit TAB here:

```
A = [1| 2
```

I would like to go to here:

```
A = [1   | 2
```

instead of indenting the current line.

**Comment by dhanak** [comment]
_(copied from a comment on #185)_

Hello!

I have set up my Emacs to upgrade my (m)elpa packages automatically on a weekly basis. Thus, it came as a surprise to me that tab completion substitution for LaTeX symbols in julia-mode broke some time earlier this week. Then it took me a while to investigate the issue and track the root cause down to #185.

Granted, I have ivy-mode installed, and I noticed the comments in the source code related to https://github.com/abo-abo/swiper/issues/2345. But that seems like an old issue, yet julia-mode worked perfectly until recently. Furthermore, there is nothing wrong with the substitution of LaTeX symbols which are unambiguous, e.g., \circeq, only those which have multiple possible suffixes, e.g., \circ and \in.

Moreover, when I type \circ<TAB> 99.99% of the time I want it to be substituted with the circle symbol, and not to be completed to some hardly ever used LaTeX symbol name that also happens to begin with \circ. The same goes for \in.

I spent the better part of an hour trying to figure out how to patch the change to bring the old behavior back, to no avail. I finally gave up and went back to the January 19 version of julia-mode which I restored from a snapshot.

Note that I don't really want to start using abbrev. As I switch back-and-forth between Emacs and REPL, I definitely want to use the same key to expand symbols, otherwise I'm sure to mess it up regularly.

I think this will be an issue for others as well, so I'd like to offer my help to come up with a solution that suits (almost) everybody.

**Comment by tpapp** [comment]
@dhanak: please read the discussion above. TL;DR: the key question is defining the semantics, after that we can implement it.

**Comment by dhanak** [comment]
I think I would prefer simple prefix completion, offering choices if there are several. In the case of your examples:

1. offer a list;
2. expand to \pre and substitute.

I don't think I have ever pressed tab in the middle of a word and expected the editor to take the suffix into account as well. When I do press tab in the middle of a word, it happens because it is not really a word, I just went back and started to enter something before what was already there. So I'd prefer the editor to leave that part as is. With a somewhat forced example:

* input: `eq|` -> `|eq` -> `not \circ|eq`, `<TAB>`, `<SPACE>`
* I expect: `not ∘ eq`
* I wouldn't like to see: `not ≗ `

**Comment by non-Jedi** [comment]
Sounds like we actually have two separate desires here:

1. Allow tab substitution while preserving indentation behavior of `(setq tab-always-indent nil)`
2. Greedily substitute on exact prefix match. Don't prompt when completing `\circ` and off choice of `\circeq`; just directly complete to `∘`.
  - As noted above, this should not apply to middle-of-symbol completion, and perhaps middle-of-symbol completion should be entirely disabled when in this mode.

Path forward I see is to have a `julia-tab-for-indent` function and a customization variable for each of the points above. `julia-tab-for-indent` would just check the values of each of the customization variables, set `tab-always-indent` and `completion-at-point-functions` to appropriate values, and then explicitly call some combination of `expand-abbrev`, `tab-for-indent`, and `completion-at-point` depending on values of customization.

**Comment by tpapp** [comment]
What we could do the following:

On `\pre|post`

1. if `pre` matches a substitution exactly, replace,
2. if not, flash choices in the Messages buffer, or popup, or say that there is absolutely no match

Otherwise, if there is no `\` in the line before the point, try indenting.

@dhanak, @ronisbr, would this match your use case?

**Comment by ronisbr** [comment]
Hi @tpapp !

> @dhanak, @ronisbr, would this match your use case?

If you mean that try indenting is calling `(indent-for-tab-command arg)`, then yes, it will match!

**Comment by dhanak** [comment]
> @dhanak, @ronisbr, would this match your use case?

Yes, that sounds good to me.

**Comment by giordano** [comment]
Do I understand correctly that at the moment there's no way to get `\alph` to be automatically completed to `\alpha` and then replaced with `α`?  I'm asking with or without pressing TAB (which would be taken care of by #196), I don't see any interactive function which can do this operation of completion+substitution at all.

**Comment by non-Jedi** [comment]
Hi @giordano. Sorry I missed your question until now. Invoking `completion-at-point` (`C-M-i` in vanilla emacs) in a julia-mode buffer will complete "\alph" directly to "α." I just tested this out using `emacs -Q` and confirmed this works for me as expected. Are you seeing an issue where this doesn't work?

**Comment by giordano** [comment]
> Invoking `completion-at-point` (`C-M-i` in vanilla emacs), in a julia-mode buffer will complete "\alph" directly to "α." I just tested this out using `emacs -Q` and installing only `julia-mode` and confirmed this works for me as expected. Are you seeing a situation where this doesn't work?

Yes, "\alph" for me is completed to "\alpha", not "α", further pressing `M-C-i` does nothing.  Not sure it's relevant, but I use Helm.  I also have auto-complete installed, but doesn't seem the mode is active in my julia buffer.

**Comment by non-Jedi** [comment]
I've managed to replicate. Will create a new issue.

**Comment by dhanak** [comment]
Eventually, I upgraded to Emacs 29, and `julia-ts-mode` with tree-sitter. Since that uses a different indentation API, this workaround didn't work for me any more, so I gave in and started to use dabbrev to auto-expand escape sequences. My main concern was that this is different from how the REPL works, and my muscle memory will surely mess it up regularly. So I tweaked the REPL instead, to work like dabbrev does. I added this code to my `startup.jl`:

```jl
atreplinit() do repl
    REPL = Base.REPL_MODULE_REF[]
    LE = REPL.LineEdit
    RC = REPL.REPLCompletions

    # define a new keymap
    my_keymap = Dict()
    push!(repl.options.extra_keymap, my_keymap)

    # auto-replace escaped latex and emoji sequences with their symbol
    # counterparts after a punctuation key, just like Emacs abbrev-mode
    function abbrev_expand(s, data, c)
        buf = LE.buffer(s)
        (pos, buf) = (position(buf), LE.content(buf))
        (comp, r, _) = RC.bslash_completions(buf, thisind(buf, pos))[2]
        if length(comp) == 1 && comp[].bslash[1] != '\\'
            LE.edit_splice!(s, prevind(buf, r.start) => r.stop, comp[].bslash)
            LE.refresh_line(s)
        end
        LE.edit_insert(s, c)
    end

    punct = filter(ispunct, Char(1):Char(127))
    for c in [' '; setdiff(punct, ";?]>")]
        my_keymap[c] = abbrev_expand
    end
end
```

**Comment by tpapp** [comment]
Just to repeat for everyone looking at this issue: #196 works so you can use it, I just did not merge it because I haven't had time to finish it properly (tests etc).

**Comment by non-Jedi** [comment]
Reopening because #196 doesn't allow you to use `TAB` for both LaTeX substitution and inserting "real" `TAB`s (see [my comment on the PR](https://github.com/JuliaEditorSupport/julia-emacs/pull/196#discussion_r1592675342)). Feel free to close again if I'm misunderstanding and the PR does enable this.

**Comment by nlw0** [comment]
I've been terribly missing tab-completions now for quite a while. `\alpha` is ok, but I can't type `\pi` without a lot of work because of silly `\pisces`. Is the PR ever getting merged? Are there any alternatives?

**Comment by dhanak** [comment]
Yes, abbrev helps. I have the following julia-mode init code in my `.emacs` file:

```elisp
(use-package julia-mode
  :hook
  (julia-mode
   . (lambda ()
       ;; backslash sequence substitution
       (abbrev-mode)
       ;; mark !, @ and _ as word constituent
       (mapc (lambda (c) (modify-syntax-entry c "w")) (string-to-list "!@_"))
       (setq-local
        paragraph-separate (concat paragraph-separate "\\|\"\"\"$")))))
```

Note that with abbrev, it is "enough" to press space to swap an escape sequence with its unicode equivalent.

**Comment by tpapp** [comment]
@nwl0: the PR is merged.

@non-Jedi: I missed the notification about the comment on the PR. I will comment soon.

**Comment by non-Jedi** [comment]
@tpapp I think you already responded here: https://github.com/JuliaEditorSupport/julia-emacs/pull/196#discussion_r1609563118

Thanks though. :)

**Comment by tpapp** [comment]
I am closing it as #196 was merged and #216 now documents how to access the feature.

**Comment by nlw0** [comment]
Finally got to update things and now I can type π again, thank you!  🙏
